092-self-healing-using-cruise-control.md

```diff
@@ -0,0 +1,273 @@
+
+Users can activate self-healing by setting the `self.healing.enabled` property to `true` and specifying a notifier that the self-healing mechanism is going to use in the `spec.cruiseControl.config` section of their `Kafka` resource.
+If the user don't set the notifier property then `NoopNotifer` will be used by default which means all the action on the anomalies would be set to `IGNORE`.
+Self-healing corresponding to broker failure or disk failure is disabled by default since we have better ways to fix these anomalies instead of moving the partition replicas across the brokers.
+For example, brokers failures are generally due to lack of resources, so we can fix the issue at cluster level and in the same way failed disks can be easily replaced with a new one, so self-healing in these cases can be more resource consuming.
+
+Here is an example of how the configured `Kafka` custom resource could look:
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    # ...
+  cruiseControl:
+    config:
+      # configuration to enable self-healing for all anomalies
+      self.healing.enabled: true
+      anomaly.notifer.class: <notifier-class> # can be set to `SelfHealingNotifier, SlackSelfHealingNotifier, AlertaSelfHealingNotifier etc. or you own custom notifier
+      self.healing.broker.failure.enabled: false # disables self healing for broker failure
+      self.healing.disk.failure.enabled: false # disables self healing for disk failure
+```
+
+The user can still enable self-healing for disk failures and broker failure if they want by setting the  `self.healing.broker.failure.enabled` for broker failure and `self.healing.disk.failure.enabled` for disk failure to `true`.
+In the same way user can enable and disable the self-healing for other anomalies.
+
+### Using metrics when using self-healing
+
+Cruise control provides [self-healing related metrics](https://github.com/linkedin/cruise-control/wiki/Sensors#anomalydetector-sensors) which can help them understand what kind of anomalies were detected in their cluster, when self-healing started, the rate at which the anomalies are being detected in their cluster and much more.
+These metrics can also provide a way to the users to debug self-healing related issues.
+For example, the `number_of_self_healing_started` metric can provides users, the way to check when self-healing was started for an anomaly, and the number of self-healing actions that were performed in the cluster. 
+This information can be useful to at what point of time the self-healing occured and the shape/replica assignment of the cluster was changed due to self-healing.
+The user can also use the  metric like `number_of_self_healing_failed` to know if the self-healing failed to start to understand, and then they can check the reasons in the log for the failure. One example could be load monitor was ready at that point of time.
```

Comment:
AFAIU all the above metrics are historical ones but they don't give any information about what's going on now.
I made this comment on the CC related PR as well, so I was wondering if we need to expose metrics about self-healing actually running.

Reply:
I don't see any exsisting metrics like that yet in CC. There are metrics which can tell that there are inter/intra broker replica movement but Iam not sure how we can expose a metric to say when self healing is running. After talking with Tom there was a way we can know if self healing still running. we can have the sum(self healing started) minus sum(self healing ended) that can tell if the self healing is still in progress

Reply:
But then it's something we should explain in the proposals and maybe providing a Grafana dashboard with such information?
Do we have, at least, the metric about when the last fix started?

Reply:
If you are asking about the existing metrics then you can still use the `number_of_self_healing_started` metric and use the promql query to get the timestamp for the last change in the gauge -> `last_over_time(timestamp(changes(kafka_cruisecontrol_anomalydetector_num_of_self_healing_started_for_broker_failure_count{}[<time-range>]) > 0`

Reply:
If we want it based on the anomaly, we can do the same with the new metrics `num_of_self_healing_started_for_<anomaly-type>_count`

Reply:
Ok so I think we should listing this info in the proposal.

Reply:
Sure

Reply:
Currently we just have single metrics called `number_of_self_healing_started` to know that self healing was started for fixing an anomaly(these metrics doesn't classify anomaly on types) and there is no metrics regarding when self healing ended. The [PR](https://github.com/linkedin/cruise-control/pull/2268) opened on CC upstream introduces new metrics regarding -> `num-of-self-healing-started-for-<anomalyType>` and ``num-of-self-healing-ended-for-<anomalyType>` which are both classified based on the anomaly types. But at the same time it doesn't add any metrics like `number_of_self_healing_ended` for self healing ended which can provide us an overall count of how many self healing ended.